File: Data Structures & Algorithms/interleaving-string/submission-1.py
class Solution:
    def recur(self, i, j, k, s1, s2, s3, dp):
        if k == len(s3):
            return i == len(s1) and j == len(s2)
        
        if (i, j) in dp:
            return dp[(i, j)]
        
        res = False
        if i < len(s1) and s1[i] == s3[k]:
            res = self.recur(i + 1, j, k + 1, s1, s2, s3, dp)
        
        if not res and j < len(s2) and s2[j] == s3[k]:
            res = self.recur(i, j + 1, k + 1, s1, s2, s3, dp)
        
        dp[(i, j)] = res
        
        return res

    def isInterleave(self, s1: str, s2: str, s3: str) -> bool:
        dp = {}
        return self.recur(0, 0, 0, s1, s2, s3, dp)
        

File: Data Structures & Algorithms/combination-target-sum/submission-1.java
class Solution {
    public List<List<Integer>> combinationSum(int[] nums, int target) {
        List<List<Integer>> ans = new ArrayList<>();
        List<Integer> tmp = new ArrayList<>();
        recur(nums.length - 1, target, nums, ans, tmp);
        return ans;
    }

    public void recur(int idx, int target, int[] nums, List<List<Integer>> ans, List<Integer> tmp) {
        if (target == 0) {
            ans.add(new ArrayList(tmp));
            return;
        }
        if (idx < 0) {
            return;
        }

        if (nums[idx] <= target) {
            tmp.add(nums[idx]);
            recur(idx, target - nums[idx], nums, ans, tmp);
            tmp.remove(tmp.size() - 1);
        }
        recur(idx - 1, target, nums, ans, tmp);
    }
}
